frontend-vue3/src/composables/api/axios.ts

```diff
@@ -0,0 +1,210 @@
+    }
+  }
+
+  async function get(url: string, options?: RequestOptions) {
+    return request(url, "GET", null, options);
+  }
+
+  async function post(url: string, data?: any, options?: RequestOptions) {
+    return request(url, "POST", data, options);
+  }
+
+  async function postForm(url: string, formData: FormData, options?: RequestOptions) {
+    return request(url, "POST", formData, options);
+  }
+
+  async function put(url: string, data: any, options?: RequestOptions) {
+    return request(url, "PUT", data, options);
+  }
+
+  async function del(url: string, options?: RequestOptions) {
+    return request(url, "DELETE", null, options);
+  }
+
+  function detailUrl(id: string | number) {
+    return `${baseUrl}${id}/`;
+  }
+
+  function searchUrl<QueryParams extends Record<string, any>>(params: QueryParams) {
+    // filter out falsy values (empty strings, etc)
+    const filteredParams: Partial<QueryParams> = {};
+    for (const key in params) {
+      if (params[key]) {
+        filteredParams[key] = params[key];
+      }
+    }
+    const qs = queryString.stringify(filteredParams);
+    if (!qs) {
+      return baseUrl || window.location.origin;
+    }
+    return `${baseUrl}?${qs}`;
+  }
+
+  return {
+    instance,
+    state,
+    get,
+    post,
+    postForm,
+    put,
+    del,
+    detailUrl,
+    searchUrl,
+  };
+}
+
+/**
+ * Utility functions
+ */
+
+function describeNonValidationError(errorResponse: AxiosResponse): string[] {
+  switch (errorResponse.status) {
+    case 403:
+      return [
+        "Server Forbidden Error (tried to read, create or modify something you do not have permission to)",
+      ];
+    case 404:
+      return [
+        "Server Resource Not Found Error (tried to read, create or modify something that does not exist)",
+      ];
+    case 500:
+      return ["Internal Server Error (server has a bug)"];
+    default:
+      return [`HTTP Error (${errorResponse.status})`];
+  }
+}
+
+function describeValidationError(errorResponse: AxiosResponse): string[] {
+  const errors = [];
+  for (const e in errorResponse.data) {
+    if (e === "non_field_errors") {
+      errors.push(...errorResponse.data[e]);
+    } else {
+      errors.push(`${e}: ${errorResponse.data[e].join(", ")}`);
+    }
+  }
+  return errors;
+}
+
+function getCookie(name: string) {
+  return document.cookie.split("; ").reduce((r, v) => {
+    const [n, ...val] = v.split("=");
+    return n === name ? decodeURIComponent(val.join("=")) : r;
+  }, "");
+}
+
+const ISODateFormat = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d*)?(?:[-+]\d{2}:?\d{2}|Z)?$/;
+
+function isISODateString(value: any) {
+  return value && typeof value === "string" && ISODateFormat.test(value);
+}
+
+function convertDates(body: any) {
+  if (body === null || body === undefined || typeof body !== "object") {
+    return body;
+  }
+  for (const key of Object.keys(body)) {
+    const value = body[key];
+    if (isISODateString(value)) {
```

Comment:
could also use `Date.parse`, `try/catch` or are we only allowing iso date strings here?

Reply:
Yeah the intent was to strictly parse ISO 8601 date strings and assume everything else is meant to be a string, the spec for `Date.parse` seemed far too loose since most implementations would parse a string like `"10 10 10"`